### DISCUSSION TITLE
Partial matching of key in setQueryData

### DISCUSSION THREAD

**Discussion by denisborovikov** [TanStack/query#394] [General] [answered]
Hi there,

I've just bumped into a weird behavior. I don't know if it's by design and expected.

Here's a minimal example https://codesandbox.io/s/sweet-goldwasser-62fhv?file=/src/App.js

We have three `useQuery` which creates three cache entities

- `["todo", { id: 1 }]`
- `["todo", { id: 2 }]`
- `["todo", { id: 3 }]`

If you click on `update cache` button in runs a cache update with `queryCache.setQueryData`. Notice that `id` there is `undefined`. In this case, all three entities will be updated, because all of them are partially matched the given key.

So my question here - is partial matching by default is a feature or a bug? In my real code when I bumped into this was that I misspelled the name of the variable `{id: data.AAA}` instead of  `{id: data.BBB`} and instead of updating the cache of the specific ID, all partially matched caches were updated (because data.AAA is undefined).

**Comment by tannerlinsley** [comment, marked as answer]
Partial matching is a feature that is baked into all of the queryCache functions that search the cache to find queries. though it may not make much sense for `setQueryData`, it does for things like `refetchQueries`.

The fact that this came by surprise to you is a good indicator that we should probably not make `setQueryData` a partial matching function.

**Reply by tannerlinsley** [reply]
The one use case I could come up with for this is if you want to set all queries in a query "family" to have the same data, but that seems like something that would never happen. Even if it was to set them all to `undefined`, I think `removeQueries` (which also supports partial matching) would be a better fit.

**Reply by denisborovikov** [reply]
I think explicit is better than implicit. It would be better to have this option to turn it on than have it always on by default.

**Reply by tannerlinsley** [reply]
Sounds good. Do you know where this logic is located? Would you like to author a PR? I can, but just want to give you the opportunity if you want it :)

**Reply by denisborovikov** [reply]
No, I don't, is it about the `exact` option?

**Reply by denisborovikov** [reply]
I think it would be better if you do it, I feel it can be pitfalls there.

**Reply by tannerlinsley** [reply]
I'll take care of it :)

**Reply by villesau** [reply]
How does it work now? I've spent a ton of time trying to understand how to do the partial match since that's what I essentially need most of the time. e.g we pull data in specific order, but we do not want to pass the order information to where we update individual items. This is not possible without partial matching.

**Reply by villesau** [reply]
There now seems to be [setQueriesData](https://react-query.tanstack.com/reference/QueryClient#queryclientsetqueriesdata) function for partial matching.

**Comment by denisborovikov** [comment]
@tannerlinsley Hey, you're about to release v2 soon. Let me remind you about this topic. Did you change the partial matching queries?

**Comment by tannerlinsley** [comment]
I will fix set query data to be exact by default in the next release.

**Reply by YassienW** [reply]
Not sure if this is the right place to ask, but is there currently a way to partially match keys when calling setQueryData?

**Reply by TkDodo** [reply]
Have a look here: https://github.com/tannerlinsley/react-query/discussions/2078#discussioncomment-618852
